src/main/scala/latis/service/hapi/DataService.scala

```diff
@@ -20,7 +19,83 @@ class DataService[F[_]: Concurrent](
   import Include._
   import QueryDecoders._
 
- private[this] val logger = getLogger
+  private[this] val logger = getLogger
+
+  private def req2Resp(r: Either[Status, DataRequest], _params:Option[NonEmptyList[String]]): F[Response[F]] = r match {
+    case Right(dataReq) => dataReq.format match {
+      case Csv =>
+        val stream = for {
+          req <- EitherT.fromEither[F](r)
+          dataset = req.dataset
+          header <- alg.getMetadata(dataset, _params).leftMap {
+            case UnknownId(_) => Status.`1406`
+            case UnknownParam(_) => Status.`1407`
+            case err@MetadataError(_) => logger.info(err.toString); Status.`1501`
+            case err@UnsupportedDataset(_) => logger.info(err.toString); Status.`1501`
+          }.map { md =>
+            "#" ++ InfoResponse(
+              HapiService.version,
+              Status.`1200`,
+              md
+            ).asJson.noSpaces
+          }
+          data <- EitherT.liftF(alg.getData(req))
+          records <- EitherT.pure[F, Status] (alg.writeCsv(data))
+          stream: Stream[F, String] = if (req.header) Stream.emit(header ++ "\n") ++ records else records
+        } yield stream
+        stream.fold(
+          {
+            case err@(Status.`1406` | Status.`1407`) =>
+              logger.info(err.message)
+              NotFound(HapiError(err).asJson)
+            case err@Status.`1501` =>
+              logger.info(err.message)
+              InternalServerError(HapiError(err).asJson)
+            case err =>
+              logger.info(err.message)
+              BadRequest(HapiError(err).asJson)
+          },
+          rs => Ok(rs).map(_.withContentType(`Content-Type`(MediaType.text.csv)))
+        ).flatten
+      case Binary =>
+        val stream = for {
+          req <- EitherT.fromEither[F](r)
+          dataset = req.dataset
+          header <- alg.getMetadata(dataset, _params).leftMap {
+            case UnknownId(_) => Status.`1406`
+            case UnknownParam(_) => Status.`1407`
+            case err@MetadataError(_) => logger.info(err.toString); Status.`1501`
+            case err@UnsupportedDataset(_) => logger.info(err.toString); Status.`1501`
+          }.map { md =>
+            "#" ++ InfoResponse(
+              HapiService.version,
+              Status.`1200`,
+              md
+            ).asJson.noSpaces
+          }
+          data <- EitherT.liftF(alg.getData(req))
+          records <- EitherT.pure[F, Status] (alg.writeBinary(data))
+          stream: Stream[F, Byte] = if (req.header) Stream.emits(header.getBytes) ++ records else records
```

Comment:
Makes sense, although I think the HAPI spec wants strings encoded to ASCII not UTF-8

Reply:
I think we should encode to UTF-8 anyway. The UTF-8 representation of ASCII characters is the same as in ASCII, so it's the same so long as we remain in the ASCII character set. We can't filter out non-ASCII characters from what we're encoding, so we might as well encode them faithfully and let reasonable clients get the actual values rather than give everyone gibberish.